### DISCUSSION TITLE
IPv6 Support

### DISCUSSION THREAD

**Discussion by scopedog** [open-webui/open-webui#2052] [General]
**Is your feature request related to a problem? Please describe.**
We are running Open WebUI in our office and need to access it from our house, too.
Since IPv6 is the only way to do so, we need IPv6 bind/listen/accept support.

**Describe the solution you'd like**
Listening to both IPv4 and IPv6 will be the best solution.

**Describe alternatives you've considered**
Adding an option for IPv6 will be also helpful.

**Additional context**
If you specify which file to modify, I'd like to do right away. Thanks for your efforts.

**Comment by justinh-rahb** [comment]
IPv6 support in Docker is spotty at best. Being that this project primarily assumes any serious usage will be deployed as a container, this is unlikely to be a high priority. Our position is that these finer points of deployment can/should be handled by the user, similar to HTTPS support, through the use of reverse-proxy servers and/or other network infrastructure.

**Comment by Infinoid** [comment]
It's true that Docker (/Podman) is stupid about IPv6 by default, but it works reasonably well once you configure it properly.  And anyway, running in a container does not necessarily imply Docker.  Kubernetes, for example, has good support for IPv6.  So I don't agree with @justinh-rahb's logic.

As the OP says, defaulting to dual-stack mode would be nice.  Failing that, it'd be helpful to be able to set the bind address to "::" via an environment variable.

This is necessary for the following use cases:
* making direct connections in dual-stack mode
* when the reverse-proxy wants to connect to the container using IPv6
* running on IPv6-native kubernetes clusters where IPv4 is disabled (yes, they really do exist)

Fortunately, the container entry script [already has a var](https://github.com/open-webui/open-webui/blob/13b0e7d64a708f69c5ce58cf0897d9951d0d16ad/backend/start.sh#L9) for this, it just isn't documented anywhere that I could find.  So the solution is simple, but it took a lot of digging to find it.

I'm running in k8s, and so I added `HOST=::` to [the relevant helm release value](https://github.com/open-webui/helm-charts/blob/3cf45141663a635cb37477eea2f7215ea8b6b524/charts/open-webui/values.yaml#L85).  Now it says:

```
INFO:     Application startup complete.
INFO:     Uvicorn running on http://[::]:8080 (Press CTRL+C to quit)
```

And it's accepting IPv6 connections just fine.

**Reply by paul8046** [reply]
This works for me too. Many thanks for putting this here - saved me hours!

**Reply by Bethibande** [reply]
Thank you. This should definitely be documented somewhere. IPv6 support is important. I just installed open-webui on a dual-stack, k3s cluster. Without this it wouldn't have been possible to get it to work, since the cluster prefers IPv6 over IPv4 and the connections where all dropped.

**Reply by AndreKR** [reply]
This also works for a regular Docker container:
```
docker run -e "HOST=::"
```

**Comment by paul8046** [comment]
The problem is that the server only binds to IPv4. Therefore, I do not agree with @justinh-rahb that the problem is the be laid at the feet of docker. While it is true that docker and podman have clunky support to IPv6 the fundamental problem here is nothing to do with them. The problem is with the webserver.

**Comment by Pleune** [comment]
Is there a way with the `--host` flag to listen dual-stack? It looks like Uvicorn supports this, but passing an empty string as the host gives an error here.

**Comment by Infinoid** [comment]
Yeah, looks like `start.sh` clobbers it in that case.  Maybe try mounting in a modified version of start.sh?

**Comment by atoomnetmarc** [comment]
I found this discussing while searching on how to enable IPv6.

I saw this in the process list:
`
/usr/local/bin/python3 -m uvicorn open_webui.main:app --host 0.0.0.0 --port 8080 --forwarded-allow-ips * --workers 1
`

Uvicorn on IPv6: https://www.uvicorn.org/settings/#socket-binding

**Comment by ljcbaby** [comment]
uvicorn need set host to None to listening dual stack.

but this option can't do by cli.

**Comment by totoCZ** [comment]
Installing on k8s gives Bad Gateway.

Fix as mentioned
```
extraEnvVars:
  - name: WEBUI_URL
    value: "https://xx"
  - name: PROXY_HEADERS
    value: "true"
  - name: HOST
    value: "::"

```
but pipelines pod stays on deprecated protocol

 /usr/local/bin/python /usr/local/bin/uvicorn main:app --host 0.0.0.0 

HOST doesn't apply.

After 1 hour Gemini made it work.

```
pipelines:
  enabled: true
  # This is the secret spot! The sub-chart looks here, not at the top level.
  extraEnvVars:
    - name: HOST
      value: "::"
```

**Comment by 24bit-xjkp** [comment]
uvicorn need set host to None to listening dual stack, in cli is `--host ""`.
But currently, `start.sh` would set `HOST` to `0.0.0.0` if the `HOST` environment variable is empty, thus only IPv4 or IPv6 can be used at once.
I think we can modify the `start.sh`, use `HOST=None` as dual stack mode, `HOST` unset as default mode and fall back to `0.0.0.0`.
